docs/specs/consensus/README.md

```diff
@@ -0,0 +1,1111 @@
+
+```text
+current:   B1, B2, B3, B4              (ye olde minimum window density = 43) Stuck! 😭
+canonical: B1, B2, B3, B4, B5, ..., Bk (current minimum window density = 42)
+```
+
+The inversion problem occurs because the calculation of the minimum window density does not take into account the relationship between the current best chain and the canonical chain with respect to time.  In Samasika, time is captured and secured through the concepts of slots and the VRF.  Our calculation of the minimum window density must also take this into account.
+
+The relative minimum window density solves this problem by projecting the joining peer's current block's window to the global slot of the candidate block. (N.b. As described in [Section 6.2.3](#623-bringing-it-all-together), this happens whenever the candidate block's global slot is ahead of the current block's or vice versa.)  In this way, the projection allows a fair comparison.
+
+The relative minimum window density of blocks `B1` and `B2` is defined as.
+
+```rust
+fn relativeMinWindowDensity(B1, B2) -> u32
+{
+    let max_slot = max(globalSlot(B1), globalSlot(B2))
+
+    // Grace-period rule
+    if max_slot < grace_period_end {
+        return B1.min_window_density
+    }
+
+    // Compute B1's window projected to max_slot
+    let projected_window = {
+        // Compute shift count
+        let shift_count = min(max(max_slot - B1.curr_global_slot - 1, 0), sub_windows_per_window)
+
+        // Initialize projected window
+        let projected_window = B1.sub_window_densities
+
+        // Ring-shift
+        let i = relativeSubWindow(B1.curr_global_slot)
+        while shift_count > 0 {
+          i = i + 1 mod sub_windows_per_window
+          projected_window[i] = 0
+          shift_count--;
+        }
+
+        return projected_window
+    }
+
+    // Compute projected window density
+    let projected_window_density = density(projected_window)
+
+    // Compute minimum window density
+    return min(B1.min_window_density, projected_window_density)
+}
+```
+
+This description was adopted to aid understanding, providing explanations where possible of why strategies have been adopted and what conditions are important.  In a production implementation consideration must be given to performance implications and another implementation may be desirable.
+
+# 6 Protocol
+
+This section specifies the consensus protocol in terms of events and how they MUST be implemented by a compatible peer.  The required events are:
+
+* [`Initialize consensus`](#61-initialize-consensus)
+* [`Select chain`](#62-select-chain)
```

Comment:
Could you please elaborate on this further and list the requirements for this RPC?

It sounds like this may belong to the P2P specification, e.g., what gossip messages are required and what RPC's are needed.

Reply:
The RPC details can live in the p2p specification, but it might be worth calling out somewhere here that consensus requires the ability to synchronize epoch ledgers from the network (for bootstrapping consensus).

Reply:
Thanks - I will add a note to call this out! 👍🏼   These area already part of the `Epoch_data` right?

Reply:
`Epoch_data` contains the commitments to the hashes of this data. The actual ledgers are stored in `Snapshot` (called epoch snapshots).

Reply:
Done